core/services/pipeline/orm.go

```diff
@@ -50,6 +58,137 @@ func (o *orm) CreateSpec(ctx context.Context, tx *gorm.DB, pipeline Pipeline, ma
+func (o *orm) StoreRun(db *sql.DB, run *Run) (restart bool, err error) {
+	finished := run.FinishedAt.Valid
+	err = postgres.SqlxTransaction(context.Background(), db, func(tx *sqlx.Tx) error {
+		if !finished {
+			// Lock the current run. This prevents races with /v2/resume
+			sql := `SELECT id FROM pipeline_runs WHERE id = $1 FOR UPDATE;`
+			if _, err = tx.Exec(sql, run.ID); err != nil {
+				return err
+			}
+
+			taskRuns := []TaskRun{}
+			// Reload task runs, we want to check for any changes while the run was ongoing
+			if err = tx.Select(&taskRuns, `SELECT * FROM pipeline_task_runs WHERE pipeline_run_id = $1`, run.ID); err != nil {
+				return err
+			}
+
+			// Construct a temporary run so we can use r.ByDotID
+			tempRun := Run{PipelineTaskRuns: taskRuns}
+
+			// Diff with current state, if updated, swap run.PipelineTaskRuns and early return with restart = true
+			for i, tr := range run.PipelineTaskRuns {
+				if !tr.IsPending() {
+					continue
+				}
+
+				// Look for new data
+				if taskRun := tempRun.ByDotID(tr.DotID); taskRun != nil {
+					// Swap in the latest state
+					run.PipelineTaskRuns[i] = *taskRun
+					restart = true
+				}
+			}
+
+			if restart {
+				return nil
+			}
+
+			// Suspend the run
+			run.State = RunStatusSuspended
+			if _, err = tx.NamedExec(`UPDATE pipeline_runs SET state = :state`, run); err != nil {
+				return err
+			}
+		} else {
+			// Simply finish the run, no need to do any sort of locking
+			if run.Outputs.Val == nil || len(run.Errors) == 0 {
+				return errors.Errorf("run must have both Outputs and Errors, got Outputs: %#v, Errors: %#v", run.Outputs.Val, run.Errors)
+			}
+			sql := `UPDATE pipeline_runs SET state = :state, finished_at = :finished_at, errors= :errors, outputs = :outputs WHERE id = :id`
+			if _, err = tx.NamedExec(sql, run); err != nil {
+				return err
+			}
+		}
+
+		sql := `
+		INSERT INTO pipeline_task_runs (pipeline_run_id, id, type, index, output, error, dot_id, created_at, finished_at)
+		VALUES (:pipeline_run_id, :id, :type, :index, :output, :error, :dot_id, :created_at, :finished_at)
+		ON CONFLICT (pipeline_run_id, dot_id) DO UPDATE SET
+		output = EXCLUDED.output, error = EXCLUDED.error, finished_at = EXCLUDED.finished_at
+		RETURNING *;
+		`
+
+		// NOTE: can't use Select() to auto scan because we're using NamedQuery,
+		// sqlx.Named + Select is possible but it's about the same amount of code
+		var rows *sqlx.Rows
+		rows, err = tx.NamedQuery(sql, run.PipelineTaskRuns)
+		if err != nil {
+			return err
+		}
+		taskRuns := []TaskRun{}
+		if err = sqlx.StructScan(rows, &taskRuns); err != nil {
+			return err
+		}
+		// replace with new task run data
+		run.PipelineTaskRuns = taskRuns
+		return nil
+	})
+	return restart, err
+}
+
+func (o *orm) UpdateTaskRun(db *sql.DB, taskID uuid.UUID, result interface{}) (run Run, start bool, err error) {
+	err = postgres.SqlxTransaction(context.Background(), db, func(tx *sqlx.Tx) error {
+		sql := `
```

Comment:
Wait don't we only want to permit this if the run is suspended? Presumably we don't want to let the resume endpoint be hit and have the task data changed for completed/non-async runs